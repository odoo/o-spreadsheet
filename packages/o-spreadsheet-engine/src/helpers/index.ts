export * from "./misc";
export * from "./numbers";
export * from "./references";
export * from "./uuid";
