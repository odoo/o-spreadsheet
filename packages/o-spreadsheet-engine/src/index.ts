export { FunctionCodeBuilder, Scope, type FunctionCode } from "./formulas/code_builder";
export {
  OPERATOR_MAP,
  UNARY_OPERATOR_MAP,
  compile,
  compileTokens,
  functionCache,
} from "./formulas/compiler";
export {
  OP_PRIORITY,
  convertAstNodes,
  iterateAstNodes,
  mapAst,
  parse,
  parseTokens,
} from "./formulas/parser";
export type {
  AST,
  ASTFuncall,
  ASTOperation,
  ASTString,
  ASTSymbol,
  ASTUnaryOperation,
} from "./formulas/parser";
export { rangeTokenize } from "./formulas/range_tokenizer";
export { POSTFIX_UNARY_OPERATORS, tokenize } from "./formulas/tokenizer";
export type { Token } from "./formulas/tokenizer";
export { FunctionRegistry } from "./functions/function_registry";
export { categories } from "./functions/functionRegistry";
export * from "./helpers";
export { Model } from "./model";
export { BasePlugin } from "./plugins/base_plugin";
export { Registry } from "./registry";
export { StateObserver, type StateObserverChange } from "./state_observer";
export * from "./translation";
export { CellValue } from "./types/cells";
export { SpreadsheetClipboardData } from "./types/clipboard";
export { CoreGetters, PluginGetters } from "./types/coreGetters";
export * from "./types/errors";
export { Format } from "./types/format";
export * from "./types/history";
export * from "./types/locale";
export * from "./types/misc";
export * from "./types/validator";
export const __info__ = {};
