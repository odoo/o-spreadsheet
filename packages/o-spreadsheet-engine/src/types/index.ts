export * from "./base";
export * from "./errors";
export * from "./functions";
export * from "./history";
export * from "./isMatrix";
export * from "./locale";
export * from "./validator";
