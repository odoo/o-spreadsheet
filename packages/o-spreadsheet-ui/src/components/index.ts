export { Spreadsheet } from "./spreadsheet/spreadsheet";
