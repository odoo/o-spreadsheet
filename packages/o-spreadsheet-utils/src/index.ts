export * from "./translation";
export * from "./uuid";
