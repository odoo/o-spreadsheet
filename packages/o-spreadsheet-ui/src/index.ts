import * as Engine from "@odoo/o-spreadsheet-engine";

export function salut() {
  console.log("Salut !");
}

export function createSpreadsheetModel() {
  return Engine.SpreadsheetModel();
}