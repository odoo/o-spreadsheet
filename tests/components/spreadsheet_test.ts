import { Component, hooks, tags } from "@odoo/owl";
import { Model } from "../../src";
import { Spreadsheet } from "../../src/components";
import { args, functionRegistry } from "../../src/functions";
import { DEBUG } from "../../src/helpers";
import { SelectionMode } from "../../src/plugins/ui/selection";
import { Client } from "../../src/types";
import { createSheet, setCellContent } from "../commands_helpers";
import { triggerMouseEvent } from "../dom_helper";
import { makeTestFixture, MockClipboard, nextTick, typeInComposer } from "../helpers";
jest.mock("../../src/components/composer/content_editable_helper", () =>
  require("./__mocks__/content_editable_helper")
);

const { xml } = tags;
const { useRef } = hooks;

let fixture: HTMLElement;
let parent: Parent;
const clipboard = new MockClipboard();

Object.defineProperty(navigator, "clipboard", {
  get() {
    return clipboard;
  },
  configurable: true,
});

class Parent extends Component<any> {
  static template = xml/* xml */ `<Spreadsheet t-ref="spreadsheet" data="data" client="client"/>`;
  static components = { Spreadsheet };
  private spreadsheet: any = useRef("spreadsheet");
  readonly data: any;
  readonly client: Client;
  get model(): Model {
    return this.spreadsheet.comp.model;
  }

  constructor(data?, client?) {
    super();
    this.data = data;
    this.client = client;
  }
}

beforeEach(async () => {
  fixture = makeTestFixture();
  parent = new Parent({
    sheets: [
      {
        id: 1,
      },
    ],
  });
  await parent.mount(fixture);
});

afterEach(() => {
  parent.destroy();
  fixture.remove();
});

describe("Spreadsheet", () => {
  test("simple rendering snapshot", async () => {
    await nextTick();
    expect(fixture.querySelector(".o-spreadsheet")).toMatchSnapshot();
  });

  test("focus is properly set, initially and after switching sheet", async () => {
    expect(document.activeElement!.tagName).toEqual("CANVAS");
    document.querySelector(".o-add-sheet")!.dispatchEvent(new Event("click"));
    // simulate the fact that a user clicking on the add sheet button will
    // move the focus to the document.body
    (document.activeElement as any).blur();
    await nextTick();
    expect(document.querySelectorAll(".o-sheet").length).toBe(2);
    expect(document.activeElement!.tagName).toEqual("CANVAS");
  });

  test("Can use the env in a function", () => {
    let env;
    functionRegistry.add("GETACTIVESHEET", {
      description: "Get the name of the current sheet",
      compute: function () {
        env = this.env;
        return "Sheet";
      },
      args: args(``),
      returns: ["STRING"],
    });
    setCellContent(parent.model, "A1", "=GETACTIVESHEET()");
    expect(env).toBeTruthy();
  });

  test("Can use the env in a function at model start", async () => {
    let env;
    functionRegistry.add("GETACTIVESHEET", {
      description: "Get the name of the current sheet",
      compute: function () {
        env = this.env;
        return "Sheet";
      },
      args: args(``),
      returns: ["STRING"],
    });
    const parent = new Parent({
      version: 2,
      sheets: [
        {
          name: "Sheet1",
          colNumber: 26,
          rowNumber: 100,
          cells: {
            A1: { content: "=GETACTIVESHEET()" },
          },
          conditionalFormats: [],
        },
      ],
      activeSheet: "Sheet1",
    });
    await parent.mount(fixture);
    expect(env).toBeTruthy();
  });

  test("Clipboard is in spreadsheet env", () => {
    expect((parent as any).spreadsheet.comp.env.clipboard).toBe(clipboard);
  });

  test("selection mode is changed with a simple select", async () => {
    expect(parent.model.getters.getSelectionMode()).toBe(SelectionMode.idle);
    triggerMouseEvent("canvas", "mousedown", 300, 200);
    expect(parent.model.getters.getSelectionMode()).toBe(SelectionMode.selecting);
    triggerMouseEvent(window, "mouseup", 300, 200);
    expect(parent.model.getters.getSelectionMode()).toBe(SelectionMode.idle);
  });

  test("selection mode is changed when selecting with CTRL pressed", async () => {
    expect(parent.model.getters.getSelectionMode()).toBe(SelectionMode.idle);
    document.activeElement!.dispatchEvent(
      new KeyboardEvent("keydown", { key: "Control", ctrlKey: true, bubbles: true })
    );
    expect(parent.model.getters.getSelectionMode()).toBe(SelectionMode.readyToExpand);
    triggerMouseEvent("canvas", "mousedown", 300, 200, { ctrlKey: true });
    expect(parent.model.getters.getSelectionMode()).toBe(SelectionMode.expanding);
    triggerMouseEvent(window, "mouseup", 300, 200, { ctrlKey: true });
    expect(parent.model.getters.getSelectionMode()).toBe(SelectionMode.readyToExpand);
    document.activeElement!.dispatchEvent(
      new KeyboardEvent("keyup", { key: "Control", bubbles: true })
    );
    expect(parent.model.getters.getSelectionMode()).toBe(SelectionMode.idle);
  });

  test("selection mode is changed when releasing CTRL while selecting", async () => {
    expect(parent.model.getters.getSelectionMode()).toBe(SelectionMode.idle);
    document.activeElement!.dispatchEvent(
      new KeyboardEvent("keydown", { key: "Control", ctrlKey: true, bubbles: true })
    );
    expect(parent.model.getters.getSelectionMode()).toBe(SelectionMode.readyToExpand);
    triggerMouseEvent("canvas", "mousedown", 300, 200, { ctrlKey: true });
    document.activeElement!.dispatchEvent(
      new KeyboardEvent("keyup", { key: "Control", bubbles: true })
    );
    expect(parent.model.getters.getSelectionMode()).toBe(SelectionMode.expanding);
    triggerMouseEvent(window, "mouseup", 300, 200);
    expect(parent.model.getters.getSelectionMode()).toBe(SelectionMode.idle);
  });

  test("selection mode is changed when pressing CTRL while selecting", async () => {
    expect(parent.model.getters.getSelectionMode()).toBe(SelectionMode.idle);
    triggerMouseEvent("canvas", "mousedown", 300, 200);
    expect(parent.model.getters.getSelectionMode()).toBe(SelectionMode.selecting);
    document.activeElement!.dispatchEvent(
      new KeyboardEvent("keydown", { key: "Control", ctrlKey: true, bubbles: true })
    );
    expect(parent.model.getters.getSelectionMode()).toBe(SelectionMode.expanding);
    triggerMouseEvent(window, "mouseup", 300, 200, { ctrlKey: true });
    expect(parent.model.getters.getSelectionMode()).toBe(SelectionMode.readyToExpand);
  });

  test("repeating CTRL keydown events does not trigger command", async () => {
    expect(parent.model.getters.getSelectionMode()).toBe(SelectionMode.idle);
    document.activeElement!.dispatchEvent(
      new KeyboardEvent("keydown", { key: "Control", ctrlKey: true, bubbles: true, repeat: true })
    );
    expect(parent.model.getters.getSelectionMode()).toBe(SelectionMode.idle);
  });

  test("Debug informations are removed when Spreadsheet is destroyed", async () => {
    parent["spreadsheet"].comp.destroy();
    expect(Object.keys(DEBUG)).toHaveLength(0);
  });

  test("can open/close search with ctrl+h", async () => {
    document.activeElement!.dispatchEvent(
      new KeyboardEvent("keydown", { key: "H", ctrlKey: true, bubbles: true })
    );
    await nextTick();
    expect(document.querySelectorAll(".o-sidePanel").length).toBe(1);
    await nextTick();
    document.activeElement!.dispatchEvent(
      new KeyboardEvent("keydown", { key: "H", ctrlKey: true, bubbles: true })
    );
    await nextTick();
    expect(document.querySelectorAll(".o-sidePanel").length).toBe(0);
  });

  test("can open/close search with ctrl+f", async () => {
    document.activeElement!.dispatchEvent(
      new KeyboardEvent("keydown", { key: "F", ctrlKey: true, bubbles: true })
    );
    await nextTick();
    expect(document.querySelectorAll(".o-sidePanel").length).toBe(1);
    await nextTick();
    document.activeElement!.dispatchEvent(
      new KeyboardEvent("keydown", { key: "F", ctrlKey: true, bubbles: true })
    );
    await nextTick();
    expect(document.querySelectorAll(".o-sidePanel").length).toBe(0);
  });

  test("Can instantiate a spreadsheet with a given client id-name", async () => {
    const client = { id: "alice", name: "Alice" };
    const parent = new Parent({}, client);
    await parent.mount(fixture);
    expect(parent.model.getters.getClient()).toEqual(client);
    parent.destroy();
  });
});

describe("Composer interactions", () => {
  test("type in grid composer adds text to topbar composer", async () => {
    document.activeElement!.dispatchEvent(
      new KeyboardEvent("keydown", { key: "Enter", bubbles: true })
    );
    await nextTick();
    const gridComposer = document.querySelector(".o-grid .o-composer");
    const topBarComposer = document.querySelector(".o-spreadsheet-topbar .o-composer");
    expect(document.activeElement).toBe(gridComposer);
    await typeInComposer(gridComposer!, "text");
    expect(topBarComposer!.textContent).toBe("text");
    expect(gridComposer!.textContent).toBe("text");
  });

  test("type in topbar composer adds text to grid composer", async () => {
    triggerMouseEvent(".o-spreadsheet-topbar .o-composer", "click");
    await nextTick();
    const topBarComposer = document.querySelector(".o-spreadsheet-topbar .o-composer");
    const gridComposer = document.querySelector(".o-grid .o-composer");
    expect(topBarComposer).toBeDefined();
    expect(document.activeElement).toBe(topBarComposer);
    expect(gridComposer).toBeDefined();
    await typeInComposer(topBarComposer!, "text");
    await nextTick();
    expect(topBarComposer!.textContent).toBe("text");
    expect(gridComposer!.textContent).toBe("text");
  });

  test("start typing in topbar composer then continue in grid composer", async () => {
    triggerMouseEvent(".o-spreadsheet-topbar .o-composer", "click");
    await nextTick();
    const topBarComposer = document.querySelector(".o-spreadsheet-topbar .o-composer");
    const gridComposer = document.querySelector(".o-grid .o-composer");

    // Type in top bar composer
    await typeInComposer(topBarComposer!, "from topbar");
    expect(topBarComposer!.textContent).toBe("from topbar");
    expect(gridComposer!.textContent).toBe("from topbar");

    // Focus grid composer and type
    triggerMouseEvent(".o-grid .o-composer", "click");
    await nextTick();
    await typeInComposer(gridComposer!, "from grid");
    expect(topBarComposer!.textContent).toBe("from topbarfrom grid");
    expect(gridComposer!.textContent).toBe("from topbarfrom grid");
  });

  test("top bar composer display active cell content", async () => {
    setCellContent(parent.model, "A2", "Hello");
    parent.model.dispatch("SELECT_CELL", { col: 0, row: 1 });
    await nextTick();
    const topBarComposer = document.querySelector(".o-spreadsheet-topbar .o-composer");
    expect(topBarComposer!.textContent).toBe("Hello");
  });

  test("autocomplete disapear when grid composer is blured", async () => {
    document.activeElement!.dispatchEvent(
      new KeyboardEvent("keydown", { key: "Enter", bubbles: true })
    );
    await nextTick();
    const topBarComposer = document.querySelector(".o-spreadsheet-topbar .o-composer")!;
    const gridComposer = document.querySelector(".o-grid .o-composer")!;
    await typeInComposer(gridComposer, "=SU");
    expect(fixture.querySelector(".o-grid .o-autocomplete-dropdown")).toBeDefined();
    topBarComposer.dispatchEvent(new Event("click"));
    await nextTick();
    expect(fixture.querySelector(".o-grid .o-autocomplete-dropdown")).toBeNull();
  });

  test("focus top bar composer does not resize grid composer when autocomplete is displayed", async () => {
    document.activeElement!.dispatchEvent(
      new KeyboardEvent("keydown", { key: "Enter", bubbles: true })
    );
    await nextTick();
    const topBarComposer = document.querySelector(".o-spreadsheet-topbar .o-composer")!;
    const gridComposerContainer = document.querySelector(".o-grid-composer")! as HTMLElement;
    const gridComposer = document.querySelector(".o-grid .o-composer")! as HTMLElement;
    const spy = jest.spyOn(gridComposerContainer.style, "width", "set");
    await typeInComposer(gridComposer, "=SU");
    await nextTick();
    topBarComposer.dispatchEvent(new Event("click"));
    await nextTick();
    expect(document.activeElement).toBe(topBarComposer);
    expect(spy).not.toHaveBeenCalled();
  });

  test("selecting ranges multiple times in topbar bar does not resize grid composer", async () => {
    triggerMouseEvent(".o-spreadsheet-topbar .o-composer", "click");
    await nextTick();
    const topBarComposer = document.querySelector(".o-spreadsheet-topbar .o-composer");
    const gridComposerContainer = document.querySelector(".o-grid-composer")! as HTMLElement;
    // Type in top bar composer
    await typeInComposer(topBarComposer!, "=");
    const spy = jest.spyOn(gridComposerContainer.style, "width", "set");
    await nextTick();
    parent.model.dispatch("SELECT_CELL", { col: 1, row: 1 });
    await nextTick();
    parent.model.dispatch("SELECT_CELL", { col: 1, row: 1 });
    await nextTick();
    expect(spy).not.toHaveBeenCalled();
  });

  test("The spreadsheet does not render after onbeforeunload", async () => {
    window.dispatchEvent(new Event("beforeunload", { bubbles: true }));
    await nextTick();
    createSheet(parent.model, {});
    await nextTick();
    const sheets = document.querySelectorAll(".o-all-sheets .o-sheet");
    expect(sheets).toHaveLength(parent.model.getters.getVisibleSheets().length - 1);
  });
});
