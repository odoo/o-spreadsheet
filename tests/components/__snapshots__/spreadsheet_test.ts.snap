// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spreadsheet simple rendering snapshot 1`] = `
<div
  class="o-spreadsheet"
>
  <div
    class="o-spreadsheet-topbar o-two-columns"
  >
    <div
      class="o-topbar-top"
    >
      <!-- Menus -->
      <div
        class="o-topbar-topleft"
      >
        <div
          class="o-topbar-menu"
          data-id="file"
        >
          File
        </div>
        <div
          class="o-topbar-menu"
          data-id="edit"
        >
          Edit
        </div>
        <div
          class="o-topbar-menu"
          data-id="view"
        >
          View
        </div>
        <div
          class="o-topbar-menu"
          data-id="insert"
        >
          Insert
        </div>
        <div
          class="o-topbar-menu"
          data-id="format"
        >
          Format
        </div>
      </div>
      <div
        class="o-topbar-topright"
      />
    </div>
    <!-- Toolbar and Cell Content -->
    <div
      class="o-topbar-toolbar"
    >
      <!-- Toolbar -->
      <div
        class="o-toolbar-tools"
      >
        <div
          class="o-tool o-disabled"
          title="Undo"
        >
          <svg
            class="o-icon"
          >
            <path
              d="M11.5656391,4.43436088 L9,7 L16,7 L16,0 L13.0418424,2.95815758 C11.5936787,1.73635959 9.72260775,1 7.67955083,1 C4.22126258,1 1.25575599,3.10984908 0,6 L2,7 C2.93658775,4.60974406 5.12943697,3.08011229 7.67955083,3 C9.14881247,3.0528747 10.4994783,3.57862053 11.5656391,4.43436088 Z"
              fill="#000000"
              transform="matrix(-1 0 0 1 17 5)"
            />
          </svg>
        </div>
        <div
          class="o-tool o-disabled"
          title="Redo"
        >
          <svg
            class="o-icon"
          >
            <path
              d="M11.5656391,4.43436088 L9,7 L16,7 L16,0 L13.0418424,2.95815758 C11.5936787,1.73635959 9.72260775,1 7.67955083,1 C4.22126258,1 1.25575599,3.10984908 0,6 L2,7 C2.93658775,4.60974406 5.12943697,3.08011229 7.67955083,3 C9.14881247,3.0528747 10.4994783,3.57862053 11.5656391,4.43436088 Z"
              fill="#000000"
              transform="translate(1 5)"
            />
          </svg>
        </div>
        <div
          class="o-tool"
          title="Paint Format"
        >
          <svg
            class="o-icon"
          >
            <path
              d="M9,0 L1,0 C0.45,0 0,0.45 0,1 L0,4 C0,4.55 0.45,5 1,5 L9,5 C9.55,5 10,4.55 10,4 L10,3 L11,3 L11,6 L4,6 L4,14 L6,14 L6,8 L13,8 L13,2 L10,2 L10,1 C10,0.45 9.55,0 9,0 Z"
              fill="#000000"
              transform="translate(3 2)"
            />
          </svg>
        </div>
        <div
          class="o-tool"
          title="Clear Format"
        >
          <svg
            class="o-icon"
          >
            <path
              d="M0.27,1.55 L5.43,6.7 L3,12 L5.5,12 L7.14,8.42 L11.73,13 L13,11.73 L1.55,0.27 L0.27,1.55 L0.27,1.55 Z M3.82,0 L5.82,2 L7.58,2 L7.03,3.21 L8.74,4.92 L10.08,2 L14,2 L14,0 L3.82,0 L3.82,0 Z"
              fill="#000000"
              transform="translate(2 3)"
            />
          </svg>
        </div>
        <div
          class="o-divider"
        />
        <div
          class="o-tool"
          title="Format as percent"
        >
          %
        </div>
        <div
          class="o-tool"
          title="Decrease decimal places"
        >
          .0
        </div>
        <div
          class="o-tool"
          title="Increase decimal places"
        >
          .00
        </div>
        <div
          class="o-tool o-dropdown"
          title="More formats"
        >
          <div
            class="o-text-icon"
          >
            123
            <svg
              class="o-icon"
            >
              <polygon
                fill="#000000"
                points="0 0 4 4 8 0"
                transform="translate(5 7)"
              />
            </svg>
          </div>
        </div>
        <div
          class="o-divider"
        />
        <!-- &lt;div class="o-tool" title="Font"&gt;&lt;span&gt;Roboto&lt;/span&gt; &lt;svg class="o-icon"&gt;&lt;polygon fill="#000000" points="0 0 4 4 8 0" transform="translate(5 7)"/&gt;&lt;/svg&gt;&lt;/div&gt; -->
        <div
          class="o-tool o-dropdown"
          title="Font Size"
        >
          <div
            class="o-text-icon"
          >
            10
             
            <svg
              class="o-icon"
            >
              <polygon
                fill="#000000"
                points="0 0 4 4 8 0"
                transform="translate(5 7)"
              />
            </svg>
          </div>
        </div>
        <div
          class="o-divider"
        />
        <div
          class="o-tool"
          title="Bold"
        >
          <svg
            class="o-icon"
          >
            <path
              d="M9,3.5 C9,1.57 7.43,0 5.5,0 L1.77635684e-15,0 L1.77635684e-15,12 L6.25,12 C8.04,12 9.5,10.54 9.5,8.75 C9.5,7.45 8.73,6.34 7.63,5.82 C8.46,5.24 9,4.38 9,3.5 Z M5,2 C5.82999992,2 6.5,2.67 6.5,3.5 C6.5,4.33 5.82999992,5 5,5 L3,5 L3,2 L5,2 Z M3,10 L3,7 L5.5,7 C6.32999992,7 7,7.67 7,8.5 C7,9.33 6.32999992,10 5.5,10 L3,10 Z"
              fill="#000000"
              fill-rule="evenodd"
              transform="translate(4 3)"
            />
          </svg>
        </div>
        <div
          class="o-tool"
          title="Italic"
        >
          <svg
            class="o-icon"
          >
            <polygon
              fill="#000000"
              fill-rule="evenodd"
              points="4 0 4 2 6.58 2 2.92 10 0 10 0 12 8 12 8 10 5.42 10 9.08 2 12 2 12 0"
              transform="translate(3 3)"
            />
          </svg>
        </div>
        <div
          class="o-tool"
          title="Strikethrough"
        >
          <svg
            class="o-icon"
          >
            <path
              d="M2.8875,3.06 C2.8875,2.6025 2.985,2.18625 3.18375,1.8075 C3.3825,1.42875 3.66,1.10625 4.02,0.84 C4.38,0.57375 4.80375,0.3675 5.29875,0.22125 C5.79375,0.075 6.33375,0 6.92625,0 C7.53375,0 8.085,0.0825 8.58,0.25125 C9.075,0.42 9.49875,0.6525 9.85125,0.95625 C10.20375,1.25625 10.47375,1.6125 10.665,2.02875 C10.85625,2.44125 10.95,2.895 10.95,3.38625 L8.6925,3.38625 C8.6925,3.1575 8.655,2.94375 8.58375,2.74875 C8.5125,2.55 8.4,2.38125 8.25,2.2425 C8.1,2.10375 7.9125,1.99125 7.6875,1.91625 C7.4625,1.8375 7.19625,1.8 6.88875,1.8 C6.5925,1.8 6.3375,1.83375 6.11625,1.8975 C5.89875,1.96125 5.71875,2.05125 5.57625,2.1675 C5.43375,2.28375 5.325,2.41875 5.25375,2.5725 C5.1825,2.72625 5.145,2.895 5.145,3.0675 C5.145,3.4275 5.32875,3.73125 5.69625,3.975 C5.71780203,3.98908066 5.73942012,4.00311728 5.76118357,4.01733315 C6.02342923,4.18863185 6.5,4.5 7,5 L4,5 C4,5 3.21375,4.37625 3.17625,4.30875 C2.985,3.9525 2.8875,3.53625 2.8875,3.06 Z M14,6 L0,6 L0,8 L7.21875,8 C7.35375,8.0525 7.51875,8.105 7.63125,8.15375 C7.90875,8.2775 8.12625,8.40875 8.28375,8.53625 C8.44125,8.6675 8.54625,8.81 8.6025,8.96 C8.65875,9.11375 8.685,9.28625 8.685,9.47375 C8.685,9.65 8.65125,9.815 8.58375,9.965 C8.51625,10.11875 8.41125,10.25 8.2725,10.35875 C8.13375,10.4675 7.95375,10.55375 7.74,10.6175 C7.5225,10.68125 7.27125,10.71125 6.97875,10.71125 C6.6525,10.71125 6.35625,10.6775 6.09,10.61375 C5.82375,10.55 5.59875,10.445 5.41125,10.3025 C5.22375,10.16 5.0775,9.9725 4.9725,9.74375 C4.8675,9.515 4.78125,9.17 4.78125,9 L2.55,9 C2.55,9.2525 2.61,9.6875 2.72625,10.025 C2.8425,10.3625 3.0075,10.66625 3.21375,10.9325 C3.42,11.19875 3.6675,11.4275 3.94875,11.6225 C4.23,11.8175 4.53375,11.9825 4.86375,12.11 C5.19375,12.24125 5.535,12.33875 5.89875,12.39875 C6.25875,12.4625 6.6225,12.4925 6.9825,12.4925 C7.5825,12.4925 8.13,12.425 8.6175,12.28625 C9.105,12.1475 9.525,11.94875 9.87,11.69375 C10.215,11.435 10.48125,11.12 10.6725,10.74125 C10.86375,10.3625 10.95375,9.935 10.95375,9.455 C10.95375,9.005 10.875,8.6 10.72125,8.24375 C10.68375,8.1575 10.6425,8.075 10.59375,7.9925 L14,8 L14,6 Z"
              fill="#010101"
              fill-rule="evenodd"
              transform="translate(2 3)"
            />
          </svg>
        </div>
        <div
          class="o-tool o-dropdown o-with-color"
        >
          <span
            style="border-color:black"
            title="Text Color"
          >
            <svg
              class="o-icon"
            >
              <path
                d="M7,0 L5,0 L0.5,12 L2.5,12 L3.62,9 L8.37,9 L9.49,12 L11.49,12 L7,0 L7,0 Z M4.38,7 L6,2.67 L7.62,7 L4.38,7 L4.38,7 Z"
                fill="#000000"
                transform="translate(3 1)"
              />
            </svg>
          </span>
        </div>
        <div
          class="o-divider"
        />
        <div
          class="o-tool o-dropdown o-with-color"
        >
          <span
            style="border-color:white"
            title="Fill Color"
          >
            <svg
              class="o-icon"
            >
              <path
                d="M14.5,8.87 C14.5,8.87 13,10.49 13,11.49 C13,12.32 13.67,12.99 14.5,12.99 C15.33,12.99 16,12.32 16,11.49 C16,10.5 14.5,8.87 14.5,8.87 L14.5,8.87 Z M12.71,6.79 L5.91,0 L4.85,1.06 L6.44,2.65 L2.29,6.79 C1.9,7.18 1.9,7.81 2.29,8.2 L6.79,12.7 C6.99,12.9 7.24,13 7.5,13 C7.76,13 8.01,12.9 8.21,12.71 L12.71,8.21 C13.1,7.82 13.1,7.18 12.71,6.79 L12.71,6.79 Z M4.21,7 L7.5,3.71 L10.79,7 L4.21,7 L4.21,7 Z"
                fill="#000000"
              />
            </svg>
          </span>
        </div>
        <div
          class="o-tool o-dropdown"
        >
          <span
            title="Borders"
          >
            <svg
              class="o-icon"
            >
              <path
                d="M0,0 L0,14 L14,14 L14,0 L0,0 L0,0 Z M6,12 L2,12 L2,8 L6,8 L6,12 L6,12 Z M6,6 L2,6 L2,2 L6,2 L6,6 L6,6 Z M12,12 L8,12 L8,8 L12,8 L12,12 L12,12 Z M12,6 L8,6 L8,2 L12,2 L12,6 L12,6 Z"
                fill="#000000"
                transform="translate(2 2)"
              />
            </svg>
          </span>
        </div>
        <div
          class="o-tool o-disabled"
          title="Merge Cells"
        >
          <svg
            class="o-icon"
          >
            <path
              d="M3,6 L1,6 L1,2 L8,2 L8,4 L3,4 L3,6 Z M10,4 L10,2 L17,2 L17,6 L15,6 L15,4 L10,4 Z M10,14 L15,14 L15,12 L17,12 L17,16 L10,16 L10,14 Z M1,12 L3,12 L3,14 L8,14 L8,16 L1,16 L1,12 Z M1,8 L5,8 L5,6 L8,9 L5,12 L5,10 L1,10 L1,8 Z M10,9 L13,6 L13,8 L17,8 L17,10 L13,10 L13,12 L10,9 Z"
              fill="#000000"
            />
          </svg>
        </div>
        <div
          class="o-divider"
        />
        <div
          class="o-tool o-dropdown"
          title="Horizontal align"
        >
          <span>
            <svg
              class="o-icon"
            >
              <path
                d="M0,14 L10,14 L10,12 L0,12 L0,14 Z M10,4 L0,4 L0,6 L10,6 L10,4 Z M0,0 L0,2 L14,2 L14,0 L0,0 Z M0,10 L14,10 L14,8 L0,8 L0,10 Z"
                fill="#000000"
                transform="translate(2 2)"
              />
            </svg>
            <svg
              class="o-icon"
            >
              <polygon
                fill="#000000"
                points="0 0 4 4 8 0"
                transform="translate(5 7)"
              />
            </svg>
          </span>
        </div>
        <!-- &lt;div class="o-tool" title="Vertical align"&gt;&lt;span&gt;&lt;svg class="o-icon"&gt;&lt;path fill="#000000" d="M9.5,3 L7,3 L7,0 L5,0 L5,3 L2.5,3 L6,6.5 L9.5,3 L9.5,3 Z M0,8 L0,10 L12,10 L12,8 L0,8 L0,8 Z M2.5,15 L5,15 L5,18 L7,18 L7,15 L9.5,15 L6,11.5 L2.5,15 L2.5,15 Z" transform="translate(3)"/&gt;&lt;/svg&gt;&lt;/span&gt; &lt;svg class="o-icon"&gt;&lt;polygon fill="#000000" points="0 0 4 4 8 0" transform="translate(5 7)"/&gt;&lt;/svg&gt;&lt;/div&gt; -->
        <!-- &lt;div class="o-tool" title="Text Wrapping"&gt;&lt;svg class="o-icon"&gt;&lt;path fill="#000000" d="M14,0 L0,0 L0,2 L14,2 L14,0 Z M0,12 L4,12 L4,10 L0,10 L0,12 Z M11.5,5 L0,5 L0,7 L11.75,7 C12.58,7 13.25,7.67 13.25,8.5 C13.25,9.33 12.58,10 11.75,10 L9,10 L9,8 L6,11 L9,14 L9,12 L11.5,12 C13.43,12 15,10.43 15,8.5 C15,6.57 13.43,5 11.5,5 Z" transform="translate(2 3)"/&gt;&lt;/svg&gt;&lt;/div&gt; -->
      </div>
      <div
        class="o-composer-container"
      >
        <div
          class="o-composer"
          contenteditable="true"
          spellcheck="false"
          style="
    line-height: 34px;
    border-bottom: 1px solid #e0e2e4;
    border-left: 1px solid #e0e2e4;
    background-color: white;
  "
          tabindex="1"
        >
          
        </div>
      </div>
    </div>
  </div>
  <div
    class="o-grid o-two-columns"
  >
    <canvas
      height="150"
      style="width:-15px;height:-15px;"
      tabindex="-1"
      width="300"
    />
    <div
      class="o-autofill"
      style="top:45px;left:140px"
    >
      <div
        class="o-autofill-handler"
        style="top:0px;left:0px;"
      />
    </div>
    <div
      class="o-overlay"
    >
      <div
        class="o-col-resizer"
      />
      <div
        class="o-row-resizer"
      />
      <div
        class="all"
      />
    </div>
    <div />
    <div
      class="o-scrollbar vertical"
    >
      <div
        style="width:1px;height:2328px"
      />
    </div>
    <div
      class="o-scrollbar horizontal"
    >
      <div
        style="height:1px;width:2592px"
      />
    </div>
  </div>
  <div
    class="o-spreadsheet-bottom-bar o-two-columns"
  >
    <div
      class="o-sheet-item o-add-sheet"
    >
      <svg
        class="o-icon"
      >
        <path
          d="M8,0 L10,0 L10,8 L18,8 L18,10 L10,10 L10,18 L8,18 L8,10 L0,10 L0,8 L8,8"
          fill="#000000"
        />
      </svg>
    </div>
    <div
      class="o-sheet-item o-list-sheets"
    >
      <svg
        class="o-icon"
        viewBox="0 0 384 384"
      >
        <rect
          height="42.667"
          width="384"
          x="0"
          y="277.333"
        />
        <rect
          height="42.667"
          width="384"
          x="0"
          y="170.667"
        />
        <rect
          height="42.667"
          width="384"
          x="0"
          y="64"
        />
      </svg>
    </div>
    <div
      class="o-all-sheets"
    >
      <div
        class="o-sheet-item o-sheet active"
        data-id="1"
        title="Sheet1"
      >
        <span
          class="o-sheet-name"
        >
          Sheet1
        </span>
        <span
          class="o-sheet-icon"
        >
          <svg
            class="o-icon"
          >
            <polygon
              fill="#000000"
              points="0 0 4 4 8 0"
              transform="translate(5 7)"
            />
          </svg>
        </span>
      </div>
    </div>
  </div>
</div>
`;
